src/com/hms_networks/americas/sc/realtimedata/RealTimeDataQueueManager.java

```diff
@@ -0,0 +1,316 @@
+      for (int i = 0; i < tagManagerD.size(); i++) {
+        numDataPoints += ((RealTimeTagDataPointManager) tagManagerD.get(i)).getSize();
+      }
+    }
+    return numDataPoints;
+  }
+
+  /**
+   * Gets a new data point for each tag in group A and adds it to an internal queue in the tag's tag
+   * Manager.
+   */
+  public static void dataFetcherA() {
+    for (int i = 0; i < tagManagerA.size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) tagsGroupA.get(i));
+      DataPoint data = getCurentTagValue(currentTag);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) tagManagerA.get(i)).addDataPoint(data);
+    }
+  }
+
+  /**
+   * Gets a new data point for each tag in group B and adds it to an internal queue in the tag's tag
+   * Manager.
+   */
+  public static void dataFetcherB() {
+    for (int i = 0; i < tagManagerB.size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) tagsGroupB.get(i));
+      DataPoint data = getCurentTagValue(currentTag);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) tagManagerB.get(i)).addDataPoint(data);
+    }
+  }
+
+  /**
+   * Gets a new data point for each tag in group C and adds it to an internal queue in the tag's tag
+   * Manager.
+   */
+  public static void dataFetcherC() {
+    for (int i = 0; i < tagManagerC.size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) tagsGroupC.get(i));
+      DataPoint data = getCurentTagValue(currentTag);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) tagManagerC.get(i)).addDataPoint(data);
+    }
+  }
+
+  /**
+   * Gets a new data point for each tag in group D and adds it to an internal queue in the tag's tag
+   * Manager.
+   */
+  public static void dataFetcherD() {
+    for (int i = 0; i < tagManagerD.size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) tagsGroupD.get(i));
+      DataPoint data = getCurentTagValue(currentTag);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) tagManagerD.get(i)).addDataPoint(data);
+    }
+  }
+
+  /**
+   * Gets a tag's current value into a list of tag current values
+   *
+   * @param tag the tag to inspect
+   * @return returns a DataPoint object holding the tags current value
+   */
+  private static DataPoint getCurentTagValue(TagInfo tag) {
+
+    String tagName = tag.getName();
+    TagType tagType = tag.getType();
+    int tagID = tag.getId();
+    String timeStampSeconds = String.valueOf(System.currentTimeMillis() / 1000);
+    DataPoint data = null;
+    TagControl tagControl = null;
+
+    try {
+      tagControl = new TagControl(tagName);
+    } catch (EWException e) {
+      e.printStackTrace();
+    }

```

Comment:
So everytime you want to read the tag you have to create a tagControl object? Seems like a ton of allocation

Reply:
Moved tag control to RealTimeDataPointManager